src/compiler/checker.ts

```diff
@@ -28733,7 +28733,17 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         // and that call signature is non-generic, return statements are contextually typed by the return type of the signature
         const signature = getContextualSignatureForFunctionLikeDeclaration(functionDecl as FunctionExpression);
         if (signature && !isResolvingReturnTypeOfSignature(signature)) {
-            return getReturnTypeOfSignature(signature);
+            const functionFlags = getFunctionFlags(functionDecl);
+
+            return filterType(getReturnTypeOfSignature(signature), returnType => {
+                if (functionFlags & FunctionFlags.Generator) {
+                    return checkGeneratorInstantiationAssignabilityToReturnType(returnType, functionFlags, /*errorNode*/ undefined);
+                }
+                if (functionFlags & FunctionFlags.Async) {
+                    return !!getAwaitedTypeOfPromise(returnType);
+                }
+                return true;
+            });

```

Comment:
I would change it to something like the following so that you don't call `filterType` when it's not needed:

```suggestion
            const functionFlags = getFunctionFlags(functionDecl);
            const returnType = getReturnTypeOfSignature(signature);
            if (functionFlags & FunctionFlags.Generator) {
                return filterType(returnType, t => checkGeneratorInstantiationAssignabilityToReturnType(t, functionFlags, /*errorNode*/ undefined));
            }
            if (functionFlags & FunctionFlags.Async) {
                return filterType(returnType, t => !!getAwaitedTypeOfPromise(returnType);
            }
            return returnType;
```